### ISSUE TITLE
Rearrange handling of stuff directory in Linux

### ISSUE THREAD

**Issue by morevnaproject** [opentoonz/opentoonz#1375]
Currently, on Linux version the "stuff" directory is located in ~/.config/OpenToonz. Its content is copied from OpenToonz distributive at the first run. It contains a lot of data, which should be outside of user directory (i.e. translations, themes, styles).

This causes problems during upgrades - i.e. if user have previous version of OpenToonz installed, then the content of stuff directory is still old. If new version introduces new translations/themes/styles(or their modifications), then user have to manually update their stuff directory.

Example of such behavior can be observed here - https://github.com/opentoonz/opentoonz/pull/1124#issuecomment-320228453

Proposed solution:
Change the way how OpenToonz searching for "stuff" data.
The default content of "stuff" directory is placed in system location -  $INSTALL_PREFIX/share/opentoonz/ (read-only).
The directory  ~/.config/OpenToonz/ is created at the start and have no data by default.

For stuff data OT crawls for two locations:
1. ~/.config/OpenToonz/
2. $INSTALL_PREFIX/share/opentoonz/

First path takes precedence over second. (It is kind of "overrides" files in read-only location).

I.e. if we have translation file in two places:
```~/.config/OpenToonz/config/loc/Deutsch/colorfx.qm```
and
```$INSTALL_PREFIX/share/opentoonz/config/loc/Deutsch/colorfx.qm```
then first one is used.

This scheme is pretty common for other applications, such as MyPaint.

**Comment by artisteacher** [comment]
Os X has the same issue when updating. I think I read on another issue that the windows installer has an option to overwrite the defaults in the stuff folder. It would be really nice to be able to update without having to figure out what needs to be deleted first.

**Comment by morevnaproject** [comment]
We have made an OSX app (Morevna Edition) and it is also affected.

>  It would be really nice to be able to update without having to figure out what needs to be deleted first.

The suggested layout with two directories  (system-space and user-space) solves that problem.

**Comment by artisteacher** [comment]
@morevnaproject, any chance this could get some attention soon? Updating the stuff folder is still quite a hassle outside of Windows.

**Comment by RodneyBaker** [comment]
I am unsure of the status of this issue.
@artisteacher as you were the last to comment on this one can you give us an update?
If still present then @morevnaproject might be able to head the problem off in time for the Linux distro.

**Comment by morevnaproject** [comment]
This problem is still present and it is unlikely that I will be able to address it before v1.4 release.

There is also another problem for Linux systems - the OpenToonz Stuff folder is located inside of `~/.config/` dir, which is hidden dir. And its meaning is to store configurations. But we have projects stored there. This is against of logic and also creates problems for users, who unable to find where their projects are stored.

**Comment by manongjohn** [comment]
While this is more problematic for Linux/macOS users, the stuff folder reorg should be applied Windows as well. 

I will post my recommendations on how the stuff folder should be split up and when it should go in a bit.

**Comment by MrBlockCat** [comment]
Maybe to go with this we could add a setting to OpenToonz's settings to change where the stuff folders are (or even just change where it looks) so we don't have to mess with files to do that, and so that can be carried over across OT versions easier? I know I'd really appreciate that!

**Comment by MrBlockCat** [comment]
Maybe when this is implemented, we could remove vanished (aka unused) translations from the code? That would make it look much neater and take up a little less space, which is always good! To my knowledge the vanished translations can be used in older versions of OpenToonz now, but it sounds like this would remove the need to keep them since translations would be built in to each version of the app?

**Comment by MrBlockCat** [comment]
Additionally, I can't speak for other operating systems, but on macOS I think it would be great if by default important OpenToonz files like translations and themes were contained within OpenToonz.app (where the rest of OpenToonz is), and files made by the user were contained in ~/Documents/OpenToonz Stuff/

**Comment by MrBlockCat** [comment]
The Windows installer not only has the overwriting option @artisteacher mentioned, but it also lets you change the location of the stuff folder! You still can't change the location OpenToonz looks at, outside of the installer, without editing the registry, but it would be awesome to get to choose the install location of the stuff folder on macOS and Linux too

**Comment by blackwarthog** [comment]
> The Windows installer not only has the overwriting option @artisteacher mentioned, but it also lets you change the location of the stuff folder! You still can't change the location OpenToonz looks at, outside of the installer, without editing the registry, but it would be awesome to get to choose the install location of the stuff folder on macOS and Linux too

i suggest to allow user to select this folder while first start
as like eclipse workspaces or firefox user profiles

We can create "Choose stuff folder" dialog will be shown before the OpenToonz appliction starts, and will contain UI to choose folder from disk. May be list of previously selected folders. And check box - "always show this dialog at startup". For people who what to active use two or more OpenToonz configurations for different purpouses. By defaut it is unchecked and dialog will not show next time you start OT.

For windows we can suggest user to choose directory in installer (as andvanced installation option). By default it will be C:/OpenToonz Stuff or C:/Users/UserName/Roamin.
After installation (and while simple installation) we do not need ask user about stuff location.

For Mac we can install stuff by default into specialized folder with application data. So we also does not need to ask user about it.

For linux i think it is bad behavior to create folders (non-hidden without dot-prefix) in user folder without a user confirmation. So i suggest to show "choose stuff folder" dialog when you run OpenToonz first time at linux. The default stuff folder location will be filled in dialog, so user can just to click "OK" to skip this dialog.

Also we need to add commandline option to force showing of "Choose stuff folder" dialog while start.
And to add such option into menu - "Restart to choose other stuff folder".

Another significant thing is upgradng of content of stuff folder.
The best way is to store system part of stuff data related with OpenToonz version inside of OpenToonz installation directory for windows and mac and in /usr/share for Linux.

But i think user can to install some extra plugins or effects into stuff folder. So we need to store info about OT version inside stuff.

While starting OpenToonz must check the version of stuff folder and show question about upgrading stuffs. If user decline upgrade then open "Choose stuff folder" to allow user to create other stuff folder.

**Comment by MrBlockCat** [comment]
I agree with I think all of that, @blackwarthog ! Maybe a button in OpenToonz's Preferences or somewhere else to pull the menu back up would be great, too, and then that could be how you turn the setting back on to show the menu when OpenToonz starts, as that checkbox would be in the menu, as in Eclipse. Though to be clear on Macs while it makes sense to by default do what you said when installing, I still do think you should be able to choose another Stuff folder location to install to if you choose, like the other OS's. That's probably what you meant too but just wanted to clarify! Also, I think that the new stuff folder should pretty much only have projects in it, and maybe other things users can add if those exist (ie. custom brushes? other things of that nature), while everything else should be contained within the OpenToonz app (ie. translations and other things that make more sense to be in the app itself)

**Comment by gab3d** [comment]
My two cents on this.
Projects folder should have by default its own directory, separated from any preferences files, that way one could do a clean install without risking the work done.

"OpenToonz Projects" dir should be in the user Home directory in each OS.
"OpenToonz Stuff" folder (whatever should be called then) should be in another place, also accessible by the user, but more "away" from normal view (in Windows, that would probably be: C:\Users\<username>\AppData\Roaming\, but I'm totally fine if it is still in C:\ where it is now, provided it is separate from the user projects dir.)

**Comment by MrBlockCat** [comment]
I can’t speak for Linux, but for macOS and Windows I know I personally would prefer the projects folder to be in the user’s Documents folder, rather than the main user/home folder. Although, for the locations of folders, even if we disagree I think what’s most important is that there’s a setting in the UI to change the folders’ locations! I agree settings and projects should be separated, though! Also, the equivalent of the AppData folder in Windows would be /Users/<username>/Library/Application Support/ in macOS

**Comment by gab3d** [comment]
yes, @MrBlockCat 
Documents folder would be the right place in Windows (it's inside the user's home dir). 😉 
Perhaps I was a little vague on that, I admit, but I didn't mean in the root of user's home, but just anywhere inside of it where it suits.

**Comment by MrBlockCat** [comment]
@RodneyBaker I'd like to add a bounty to this issue, as it's very important to me! Is the issue/goal clear enough that that would be doable? (hopefully you're the right person to ask about that) Thanks!

**Comment by RodneyBaker** [comment]
It's not entirely clear to me what the proposed outcome is but perhaps part of the bounty is for R&D to determine what that outcome optimally should be.

From a quick read it appears that Linux and Mac specifically have issues with the placement of the stuff directory... as it is problematic in Opentoonz functioning... while it is less of a concern with Windows.

**Comment by morevnaproject** [comment]
> Projects folder should have by default its own directory, separated from any preferences files, that way one could do a clean install without risking the work done.

I am absolutely agree with @gab3d.

Although, the scope of my issue initially was different, we can start the ball rolling by changing location of projects dir.

On the other hand, we can fully eliminate the need for defining project dir if https://github.com/opentoonz/opentoonz/issues/3259 will be implemented. ^__^

**Comment by RodneyBaker** [comment]
Closing this as likely resolved although unconfirmed.

Recommend opening a new report with examples, video, project files in order to lock the issue down and target it for resolution.

Recommend referencing this issue in any newly opened report.
